### PR TITLE
Fix: 2024/04/18 update

### LINKED ISSUE #20: connect.ext - Failed to find CBaseClient__SetSteamID function.
L 04/19/2024 - 11:05:36: [SM] Unable to load extension "connect.ext": Failed to find CBaseClient__SetSteamID function.

**Possible signature update required?**  for CBaseClient__SetSteamID

[connect.games.txt](https://github.com/asherkin/connect/files/15043993/connect.games.txt)

Windows TF2 server
Latest SM build 1.12.0.7116

It was working perfectly before the update of 18/04/2024.

**I have attached the files for reference.**

```
"Games"
{
	"#default"
	{
		"#supported"
		{
			"engine"    "dods"
			"engine"    "css"
			"engine"    "hl2dm"
			"engine"    "tf2"
		}

		"Offsets"
		{
			"ISteamGameServer__BeginAuthSession"
			{
				"linux"         "29"
				"mac"           "29"
				"windows"       "29"
			}

			"ISteamGameServer__EndAuthSession"
			{
				"linux"         "30"
				"mac"           "30"
				"windows"       "30"
			}
		}

		"Signatures"
		{
			"CBaseServer__CheckMasterServerRequestRestart"
			{
				"library"       "engine"
				"windows"       "\xE8\x2A\x2A\x2A\x2A\x83\x78\x04\x00\x74\x2A\xE8\x2A\x2A\x2A\x2A\x8B\x48\x04\x8B\x01\x8B\x40\x2C\xFF\xD0\x84\xC0\x74\x2A\x56"
			}

			"Steam3Server"
			{
				"library"       "engine"
				"linux"         "@_Z12Steam3Serverv"
				"mac"           "@_Z12Steam3Serverv"
			}
		}
	}

	"#default"
	{
		"#supported"
		{
			"engine"    "dods"
			"engine"    "css"
			"engine"    "hl2dm"
		}

		"Signatures"
		{
			"CBaseServer__ConnectClient"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer13ConnectClientER8netadr_siiiiPKcS3_S3_i"
				"mac"           "@_ZN11CBaseServer13ConnectClientER8netadr_siiiiPKcS3_S3_i"
				"windows"       "\x55\x8B\xEC\x81\xEC\x04\x05\x00\x00\x56\x68\x2A\x2A\x2A\x2A\x8B\xF1"
			}

			"CBaseServer__RejectConnection"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer16RejectConnectionERK8netadr_siPKc"
				"mac"           "@_ZN11CBaseServer16RejectConnectionERK8netadr_siPKc"
				"windows"       "\x55\x8B\xEC\x81\xEC\x04\x05\x00\x00\x56\x6A\xFF"
			}

			"CBaseClient__SetSteamID"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseClient10SetSteamIDERK8CSteamID"
				"mac"           "@_ZN11CBaseClient10SetSteamIDERK8CSteamID"
				"windows"       "\x55\x8B\xEC\x8B\x55\x08\x8B\x02\x89\x41\x59\x8B\x42\x04"
			}

			"CBaseServer__CheckChallengeType"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
				"mac"           "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
				"windows"       "\x55\x8B\xEC\x83\xEC\x14\x56\x57\x8B\x7D\x2A\x8B\xF1"
			}
		}
	}

	"#default"
	{
		"#supported"
		{
			"engine"    "tf2"
		}

		"Signatures"
		{
			"CBaseServer__ConnectClient"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer13ConnectClientER8netadr_siiiiPKcS3_S3_i"
				"mac"           "@_ZN11CBaseServer13ConnectClientER8netadr_siiiiPKcS3_S3_i"
				"windows"       "\x55\x8B\xEC\x81\xEC\x24\x05\x00\x00"
			}

			"CBaseServer__RejectConnection"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer16RejectConnectionERK8netadr_siPKc"
				"mac"           "@_ZN11CBaseServer16RejectConnectionERK8netadr_siPKc"
				"windows"       "\x55\x8B\xEC\x81\xEC\x04\x05\x00\x00\x57"
			}

			"CBaseClient__SetSteamID"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseClient10SetSteamIDERK8CSteamID"
				"mac"           "@_ZN11CBaseClient10SetSteamIDERK8CSteamID"
				"windows"       "\x55\x8B\xEC\x56\x57\x8B\x7D\x08\x8B\xF1\x8D\x4E\x04"
			}

			"CBaseServer__CheckChallengeType"
			{
				"library"       "engine"
				"linux"         "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
				"mac"           "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
				"windows"       "\x55\x8B\xEC\x83\xEC\x14\x53\x8B\x5D\x14"
			}
		}
	}
}
```

### PR DESCRIPTION
**Pull request by Natanel-Shitrit**
Resolves #20

### CODE REVIEW

File: extension/extension.cpp [hunk 1]
```diff
@@ -342,8 +342,15 @@ bool Connect::SDK_OnLoad(char *error, size_t maxlen, bool late)
 		return false;
 	}
 
+	int steam3ServerFuncOffset = 0;
```
Comment (sapphonie) [inline extension/extension.cpp:345]: should be intptr_t ☝🤓

Comment (Natanel-Shitrit) [reply extension/extension.cpp:345]: I'm following the code convention ☝🤓
https://github.com/asherkin/connect/blob/0c6274058e718d142189d331aae9b288b351360a/extension/extension.cpp#L149-L150

Comment (asherkin) [reply extension/extension.cpp:345]: Admittedly those ones are used as array indices, but I'm happy with either.

File: connect.games.txt [hunk 2]
```diff
@@ -1,134 +1,139 @@
...
+			}
+		}
+
+		"Signatures"
+		{
+			"CBaseServer__CheckMasterServerRequestRestart"
+			{
+				"library"       "engine"
+				"windows"       "\x55\x8B\xEC\x83\xEC\x1C\x53\x57\x33\xD2"
+			}
```
Comment (asherkin) [inline connect.games.txt:41]: The old one here was a midfunc sig to the right bit of the function, the new sig+offset is definitely cleaner though.

File: connect.games.txt [hunk 3]
```diff
@@ -1,134 +1,139 @@
...
+			"ISteamGameServer__EndAuthSession"
+			{
+				"linux"         "30"
+				"mac"           "30"
+				"windows"       "30"
+			}
+
+			"CheckMasterServerRequestRestart_Steam3ServerFuncOffset"
+			{
+				"windows"		"240"
```
Comment (asherkin) [inline connect.games.txt:31]: Is this the same for dods, css, and hl2dm?

Comment (Natanel-Shitrit) [reply connect.games.txt:31]: Currently only tested for TF2. (by @khalik38 + @caxanga334)
I hope someone is willing to test :D

File: connect.games.txt (hunk 2 above)
Comment (Natanel-Shitrit) [reply connect.games.txt:41]: I just realized that now 😄
Totally missed that!

File: connect.games.txt (hunk 3 above)
Comment (Natanel-Shitrit) [reply connect.games.txt:31]: The best I can do is check the binaries.
I'll update soon

Comment (Natanel-Shitrit) [reply connect.games.txt:31]: dods binaries are out-of-date 😢 
https://users.alliedmods.net/~asherkin/public/bins/source/dods/

File: connect.games.txt [hunk 4]
```diff
@@ -1,134 +1,139 @@
...
-			{
-				"library"       "engine"
-				"linux"         "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
-				"mac"           "@_ZN11CBaseServer18CheckChallengeTypeEP11CBaseClientiR8netadr_siPKcii"
-				"windows"       "\x55\x8B\xEC\x83\xEC\x14\x53\x8B\x5D\x14"
-			}
-		}
-	}
-}
+"Games"
```
Comment (sapphonie) [inline connect.games.txt:1]: Can you please fix the whitespace so it does not generate a diff 🙏

**Comment by sapphonie** [comment]
Does this resolve #19 ?

**Comment by Natanel-Shitrit** [comment]
shit

**Comment by Natanel-Shitrit** [comment]
This is currently crashing upon client connection (at least according to @khalik38)
Anyone here willing to help with testing on a server with [Accelerator](https://forums.alliedmods.net/showthread.php?t=277703) installed? (and share with me stack, etc...)

**Comment by leighmacdonald** [comment]
> This is currently crashing upon client connection (at least according to @khalik38) Anyone here willing to help with testing on a server with [Accelerator](https://forums.alliedmods.net/showthread.php?t=277703) installed? (and share with me stack trace, etc...)

@Natanel-Shitrit if you have a linux build sure

**Comment by Natanel-Shitrit** [comment]
@leighmacdonald Let me add artifacts feature to the CI/CD so we can download directly from here

**Comment by Dragonisser** [comment]
Would gladly test as well when a linux binary is available.

**Comment by Natanel-Shitrit** [comment]
This PR does not change anything for the Linux build.
gamedata is Windows only and so as the midfunc change.

Latest build available [here](https://builds.limetech.io/?p=connect) (or rather you have it installed already)

**Comment by Dragonisser** [comment]
> This PR does not change anything for the Linux build. gamedata is Windows only and so as the midfunc change.
> 
> Latest build available [here](https://builds.limetech.io/?p=connect) (or rather you have it installed already)

oh, whoops. I assumed it was in some form related to #19 🙃

**Comment by leighmacdonald** [comment]
> > This PR does not change anything for the Linux build. gamedata is Windows only and so as the midfunc change.
> > Latest build available [here](https://builds.limetech.io/?p=connect) (or rather you have it installed already)
> 
> oh, whoops. I assumed it was in some form related to #19 🙃

Whoops, yeah, should have looked at the actual commits, thought the same. :+1:

**Comment by khalik38** [comment]
Can someone compile the extension so it can be downloaded and tested.
I am unable to compile the extension.

**Comment by Natanel-Shitrit** [comment]
> Can someone compile the extension so it can be downloaded and tested. I am unable to compile the extension.

The [version](https://discord.com/channels/335290997317697536/336023718709035020/1230985827450884127) @caxanga334 compiled in the AM discord is up-to-date.
I just need access to the full information of the crash dump 😄

**Comment by Natanel-Shitrit** [comment]
cmon git 😮‍💨

**Comment by khalik38** [comment]
How do I download the compiled binary extension?  any link

**Comment by Natanel-Shitrit** [comment]
Seems to be working according to @khalik38! 😄
@sapphonie If you think single confirmation is enough then this is ready to be merged (just need to fix the git fuckery)

**Comment by khalik38** [comment]
[connect.ext.tf2.zip](https://github.com/asherkin/connect/files/15068952/connect.ext.tf2.zip)

I can confirm it is working on TF2 Windows.

Extension loaded.
         [03] Connect (1.2.0-manual): Forward for early connection

             Metamod:Source version 2.0.0-dev+1289
             SourceMod Version: 1.12.0.7121

**Comment by Natanel-Shitrit** [comment]
Okay git all fixed up, I've split the offset and signature since it's different between TF2 v. non-TF2.

Original offset was 1, I've checked CSS `engine.dll` and indeed it was correct, psychonic said all other non-TF2 games have the same engine build so it should be fine 😄

@sapphonie sorry for tagging spam, it's all ready for merge now.

**Comment by Natanel-Shitrit** [comment]
> > Okay git all fixed up, I've split the offset and signature since it's different between TF2 v. non-TF2.
> > 
> > Original offset was 1, I've checked CSS `engine.dll` and indeed it was correct, psychonic said all other non-TF2 games have the same engine build so it should be fine 😄
> > 
> > @sapphonie sorry for tagging spam, it's all ready for merge now.
> 
> Just to clarify, this doesn't resolve #19, correct?
> 

No, this only changes Windows stuff.

**Comment by sapphonie** [comment]
> > > Okay git all fixed up, I've split the offset and signature since it's different between TF2 v. non-TF2.
> > > 
> > > Original offset was 1, I've checked CSS `engine.dll` and indeed it was correct, psychonic said all other non-TF2 games have the same engine build so it should be fine 😄
> > > 
> > > @sapphonie sorry for tagging spam, it's all ready for merge now.
> > 
> > Just to clarify, this doesn't resolve #19, correct?
> > 
> 
> No, this only changes Windows stuff.

Ty. Ill merge this. Hopefully someone steps up 4 the linux issue because I am unable to with work atm

**Pull request merged**